src/api/gradient.jl

```diff
@@ -59,11 +83,51 @@ function gradient{A}(f, ::Type{A}=Void;
     end
 end
 
+function gradient{A, T}(f, input_size::Int, input_type::Type{T},
```

Comment:
This whole 30 line block here is basically only to use a `@genereated` function to compute the type of the GradientCache (to avoid type instability) and create the cache...